src/behaviors/focusTrap.ts

```diff
@@ -64,55 +64,51 @@ export function focusTrap(
 
   container.setAttribute('data-focus-trap', 'active')
   let lastFocusedChild: HTMLElement | undefined = undefined
-  const firstFocusableChild = getFocusableChild(container)
-  const lastFocusableChild = getFocusableChild(container, true)
 
   // Ensure focus remains in the trap zone by checking that a given recently-focused
   // element is inside the trap zone. If it isn't, redirect focus to a suitable
   // element within the trap zone. If need to redirect focus and a suitable element
   // is not found, return false.
-  function ensureTrapZoneHasFocus(focusedElement: EventTarget | null): boolean {
+  function ensureTrapZoneHasFocus(focusedElement: EventTarget | null) {
     if (focusedElement instanceof HTMLElement) {
       if (container.contains(focusedElement)) {
         // If a child of the trap zone was focused, remember it
         lastFocusedChild = focusedElement
-        return true
+        return
       } else {
-        if (lastFocusedChild && isTabbable(lastFocusedChild)) {
+        if (lastFocusedChild && isTabbable(lastFocusedChild) && container.contains(lastFocusedChild)) {
           lastFocusedChild.focus()
-          return true
+          return
         } else {
-          // eslint-disable-next-line @typescript-eslint/no-extra-semi
-          const toFocus = initialFocus ?? firstFocusableChild
+          const toFocus = initialFocus && container.contains(initialFocus) ? initialFocus : getFocusableChild(container)
           if (toFocus) {
             toFocus.focus()
-            return true
+            return
+          } else {
+            // no element focusable within trap, blur the external element instead
+            focusedElement.blur()
           }
         }
       }
     }
-    return false
   }
 
   const wrappingController = followSignal(signal)
 
-  // Wrap focus from beginning to end
-  firstFocusableChild?.addEventListener(
+  container.addEventListener(
     'keydown',
     event => {
-      if (event.key === 'Tab' && event.shiftKey && !event.defaultPrevented) {
-        event.preventDefault()
-        lastFocusableChild?.focus()
+      if (event.key !== 'Tab' || event.defaultPrevented) {
+        return
       }
-    },
-    {signal: wrappingController.signal}
-  )
 
-  // Wrap focus from end to beginning
-  lastFocusableChild?.addEventListener(
-    'keydown',
-    event => {
-      if (event.key === 'Tab' && !event.shiftKey && !event.defaultPrevented) {
+      const {target} = event
+      const firstFocusableChild = getFocusableChild(container)
```

Comment:
I was thinking this would not be great for perf, since it would cause a reflow every tab key press (including rapid succession, like holding down tab), but then I realized [focus is going to cause a reflow anyway](https://source.chromium.org/chromium/chromium/src/+/master:third_party/blink/renderer/core/dom/element.cc;l=4206-4225;drc=d685ea3c9ffcb18c781bc3a0bdbb92eb88842b1b) :)